context/app/static/js/components/workspaces/hooks.js

```diff
@@ -1,61 +1,78 @@
+import { useContext } from 'react';
+import useSWR from 'swr';
 
 import { AppContext } from 'js/components/Providers';
 
-import { mergeJobsIntoWorkspaces, createWorkspaceAndNotebook, deleteWorkspace, stopJobs } from './utils';
+import { mergeJobsIntoWorkspaces, createWorkspaceAndNotebook, deleteWorkspace, stopJobs, startJob } from './utils';
 
-function useWorkspacesList() {
-  const [workspacesList, setWorkspacesList] = useState([]);
-  // TODO: isLoading:
-  // const [isLoading, setIsLoading] = useState(true);
+async function fetchWorkspaces(workspacesEndpoint, workspacesToken) {
+  const fetchOpts = {
+    method: 'GET',
+    headers: {
+      'Content-Type': 'application/json',
+      'UWS-Authorization': `Token ${workspacesToken}`,
+    },
+  };
 
-  const { workspacesEndpoint, workspacesToken } = useContext(AppContext);
-  const getAndSetWorkspaces = useCallback(async () => {
-    const fetchOpts = {
-      method: 'GET',
-      headers: {
-        'Content-Type': 'application/json',
-        'UWS-Authorization': `Token ${workspacesToken}`,
-      },
-    };
-    const [workspacesResponse, jobsResponse] = await Promise.all([
-      fetch(`${workspacesEndpoint}/workspaces`, fetchOpts),
-      fetch(`${workspacesEndpoint}/jobs`, fetchOpts),
-    ]);
-
-    if (!workspacesResponse.ok || !jobsResponse.ok) {
-      console.error('Workspaces API failed. Workspaces:', workspacesResponse, 'Jobs:', jobsResponse);
-      return;
-    }
+  const [workspacesResponse, jobsResponse] = await Promise.all([
+    fetch(`${workspacesEndpoint}/workspaces`, fetchOpts),
+    fetch(`${workspacesEndpoint}/jobs`, fetchOpts),
+  ]);
+
+  if (!workspacesResponse.ok || !jobsResponse.ok) {
+    console.error('Workspaces API failed. Workspaces:', workspacesResponse, 'Jobs:', jobsResponse);
+  }
 
-    // This is could be parallelized too...
-    // but since it's not network-bound, little benefit.
-    const workspaceResults = await workspacesResponse.json();
-    const jobsResults = await jobsResponse.json();
+  // This is could be parallelized too...
+  // but since it's not network-bound, little benefit.
+  const workspaceResults = await workspacesResponse.json();
+  const jobsResults = await jobsResponse.json();
 
-    const workspaces = mergeJobsIntoWorkspaces(jobsResults.data.jobs, workspaceResults.data.workspaces);
+  return mergeJobsIntoWorkspaces(jobsResults.data.jobs, workspaceResults.data.workspaces);
+}
 
-    setWorkspacesList(workspaces);
-  }, [workspacesEndpoint, workspacesToken]);
+function useWorkspacesList() {
+  const { workspacesEndpoint, workspacesToken } = useContext(AppContext);
+
+  const { data: workspacesList, mutate } = useSWR(
+    ['workspaces', workspacesToken],
+    ([, token]) => fetchWorkspaces(workspacesEndpoint, token),
+    { fallbackData: [] },
+  );
 
   async function handleDeleteWorkspace(workspaceId) {
     await deleteWorkspace({ workspaceId, workspacesEndpoint, workspacesToken });
-    getAndSetWorkspaces();
+    mutate();
```

Comment:
If we know how each of these actions will affect the local store of workspaces, we could optimistically update the local data and reflect the change a tiny bit faster, e.g. https://swr.vercel.app/docs/mutation#bound-mutate

The optimistic update would still be reconciled with the server data once it's received, but if there are any UI elements that are visible to the user when this fires, then the update would be reflected as soon as the awaited action before it returns instead of needing to wait for another round trip to the servers, and would appear that much more responsive to users.

Reply:
Thanks! The workspaces UI is very much in beta and we expect it to change substantially in the coming months. I don't want to over-optimize while it's in this state of flux, but I hope to introduce optimistic updates in the future.